### PR TITLE
fix: MonoTypes enum erasure

### LINKED ISSUE #7173: Erase enum type arguments properly
MonoTyper visitType

and make sure that verifier accepts it

### PR DESCRIPTION
**Pull request by JonathanStarup**
Fixes #7173

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@mlutze I see some kinding stuff going, so maybe you can also take a look?

File: main/src/ca/uwaterloo/flix/language/phase/MonoTypes.scala [hunk 1]
```diff
@@ -290,25 +293,72 @@ object MonoTypes {
...
-      // Assumed to have been removed earlier.
-      throw InternalCompilerException(s"Unexpected type var: '$sym'", loc)
-    case Type.AssocType(cst, _, _, loc) =>
-      // Assumed to have been removed earlier.
-      throw InternalCompilerException(s"Unexpected associated type: '${cst.sym}'", loc)
+  @tailrec
+  private def kindArity(kind0: Kind, acc: Int, loc: SourceLocation): Int = {
+    kind0 match {
+      case Kind.Star => acc
+      case Kind.Arrow(_, k2) => kindArity(k2, acc+1, loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/MonoTypes.scala:359]: No space between -? Forgot to use IDEA formatter? :)

**Comment by magnus-madsen** [comment]
Would the following simpler (but more expensive) algorithm be obviously correct:

- Match on tpe.typeConstructor.
- If enum then simply return that enum.
- Otherwise recurse, rebuilding the type.

Downside: Nˆ2 complexity, but not a problem, because types are very small.

**Comment by magnus-madsen** [comment]
@JonathanStarup After this, merge the phases?

**Comment by JonathanStarup** [comment]
> Would the following simpler (but more expensive) algorithm be obviously correct:
> 
>     * Match on tpe.typeConstructor.
> 
>     * If enum then simply return that enum.
> 
>     * Otherwise recurse, rebuilding the type.
> 
> 
> Downside: Nˆ2 complexity, but not a problem, because types are very small.

Yeah, that would work 👍

**Comment by magnus-madsen** [comment]
> > Would the following simpler (but more expensive) algorithm be obviously correct:
> > ```
> > * Match on tpe.typeConstructor.
> > 
> > * If enum then simply return that enum.
> > 
> > * Otherwise recurse, rebuilding the type.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Downside: Nˆ2 complexity, but not a problem, because types are very small.
> 
> Yeah, that would work 👍

Lets do it.

**Comment by magnus-madsen** [comment]
Better yes? RFR?

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
After this, look into merging the phases?

**Comment by JonathanStarup** [comment]
ready to merge on my account yes

**Comment by magnus-madsen** [comment]
Next, drop MonoTypes phase? 🎆 ?

**Pull request merged**